{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2f8da833-d329-4252-a7df-b6e002f485b4",
   "metadata": {},
   "source": [
    "## Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "448462fd-7299-492d-8a8d-d78b34ad103e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Helper_functions import Get_valid_chunk_sizes\n",
    "from Helper_functions import Build_canonical_map\n",
    "from Helper_functions import Get_seed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1f4ea399-36c8-4d8d-a258-a423c8d281f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import product \n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "from scipy.interpolate import PchipInterpolator\n",
    "from pathlib import Path\n",
    "import joblib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "64f77eef-d9e4-4e40-9331-7d4ff8f37000",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Generate_arrays(Valid_chunk_sizes, Kmer_size, Canonical_map, Chromosome_seq, Chromosome_id, Num_chunks):\n",
    "    \n",
    "    # generate a kmer --> index for fast numpy indexing \n",
    "    All_kmers = sorted(set(Canonical_map.values()))\n",
    "    Kmer_index = {kmer : index for index, kmer in enumerate(All_kmers)}\n",
    "\n",
    "    Counts_list = []\n",
    "    Cov_list = []\n",
    "\n",
    "    for Chunk_size in Valid_chunk_sizes:\n",
    "        #Numpy array for storing kmers counts at that given chunk_size \n",
    "        Counted_chunks_array = np.zeros((Num_chunks, len(All_kmers)), dtype=\"int64\") # shape = (N_chunks, N_kmers)\n",
    "\n",
    "\n",
    "        Half_chunk_size = Chunk_size // 2\n",
    "        for Chunk in range(Num_chunks):\n",
    "            # generate a random seed location to cut the chromosome at \n",
    "            Seed_location = Get_seed(Chromosome_id, Half_chunk_size, Chunk, len(Chromosome_seq), Interpolator = True)\n",
    "            # Count the kmers across that chunk, use numpy indexing (fast) to input into array \n",
    "            Start, Stop = int(Seed_location - Half_chunk_size), int(Seed_location + Half_chunk_size - Kmer_size + 1)\n",
    "            for Window in range(Start, Stop):\n",
    "                Raw_kmer = Chromosome_seq[Window : (Window + Kmer_size)]\n",
    "                Canonical_kmer = Canonical_map.get(Raw_kmer)\n",
    "                # Input into array \n",
    "                col = Kmer_index[Canonical_kmer]\n",
    "                Counted_chunks_array[Chunk, col] += 1  \n",
    "                \n",
    "        # Calculte the mean kmer counts across all of the chunks \n",
    "        Mean_count = Counted_chunks_array.mean(axis = 0)\n",
    "        # add to List \n",
    "        Counts_list.append(Mean_count)\n",
    "        \n",
    "        # Calculate the Cov --> Creates a 136 by 136 matrix \n",
    "        Mean_cov = np.cov(Counted_chunks_array, rowvar = False)\n",
    "        # add to list\n",
    "        Cov_list.append(Mean_cov)\n",
    "    \n",
    "    return Counts_list, Cov_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dd4d7c96-fc10-4243-bf33-73af8d48ee48",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def Generate_counts_interpolator(Mean_counts_list, Valid_chunk_sizes):\n",
    "    #convert into array \n",
    "    chunk_sizes_array = np.array(Valid_chunk_sizes)\n",
    "    mean_vecs_array = np.asarray(Mean_counts_list) # Shape (n_chunk_sizes, n_kmers)\n",
    "    n_chunk_sizes, n_kmers = mean_vecs_array.shape\n",
    "\n",
    "    mean_vec_interpolators = []\n",
    "\n",
    "    # iterate through each of the kmers \n",
    "    for kmer in range(n_kmers):\n",
    "        # get all of the values for that kmer across all of the arrays \n",
    "        values_at_kmer = mean_vecs_array[:, kmer] \n",
    "        # interpolate across those values \n",
    "        interpolator = PchipInterpolator(chunk_sizes_array, values_at_kmer, extrapolate = False)\n",
    "        # Add interpolator function into list \n",
    "        mean_vec_interpolators.append(interpolator)\n",
    "   \n",
    "    return mean_vec_interpolators \n",
    "    \n",
    "\n",
    "def Generate_cov_interpolator(Mean_cov_list, Valid_chunk_sizes):\n",
    "\n",
    "    # convert into array \n",
    "    cov_vecs_array = np.asarray(Mean_cov_list) # shape = (10, 136, 136)\n",
    "    chunk_sizes_array = np.array(Valid_chunk_sizes)\n",
    "\n",
    "    n_chunk_sizes, kmer_index, kmer_columns = cov_vecs_array.shape\n",
    "\n",
    "    cov_interpolators = []\n",
    "\n",
    "    for kmer1 in range(kmer_index):\n",
    "        row_interpolators = []\n",
    "        for kmer2 in range(kmer_columns):\n",
    "            # retrieve the values of kmer 1 + kmer 2 pairings across all of the arrays \n",
    "            values_at_pairing = cov_vecs_array[:, kmer1, kmer2]\n",
    "            interpolator = PchipInterpolator(chunk_sizes_array, values_at_pairing, extrapolate = False)\n",
    "            row_interpolators.append(interpolator)\n",
    "        cov_interpolators.append(row_interpolators)\n",
    "    return cov_interpolators\n",
    "\n",
    "\n",
    "def Export_Interpolators(Mean_counts_interpolator, Cov_interpolator, Valid_chunk_sizes, Kmer_size, Chromosome_id, Genome_name, Output_dir):\n",
    "\n",
    "    Output_path = Path(Output_dir, Genome_name)\n",
    "    Output_path.mkdir(parents = True, exist_ok = True)\n",
    "    Output_file = Output_path / f\"{Chromosome_id}.pkl\"\n",
    "\n",
    "    Bundle = {\n",
    "        \"Mean_counts_interpolator\" : Mean_counts_interpolator,\n",
    "        \"Cov_interpolator\" : Cov_interpolator,\n",
    "        \"Chunk_sizes\" : Valid_chunk_sizes,\n",
    "        \"Kmer_size\" : Kmer_size,\n",
    "        \"Meta_data\" : {\"Genome_name\" : Genome_name, \"Chromosome_id\" : Chromosome_id}\n",
    "                        \n",
    "    }\n",
    "    joblib.dump(Bundle, Output_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3a536e19-0f07-4186-b2db-b28a1ef635de",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing Genome : 1Deinococcus_radiodurans\n",
      "Processing Chromosome : NZ_CP038663.1\n",
      "Saving Chromosome : NZ_CP038663.1\n",
      "Processing Chromosome : NZ_CP038664.1\n",
      "Saving Chromosome : NZ_CP038664.1\n",
      "Processing Chromosome : NZ_CP038666.1\n",
      "Saving Chromosome : NZ_CP038666.1\n",
      "Processing Chromosome : NZ_CP038665.1\n",
      "Saving Chromosome : NZ_CP038665.1\n",
      "Processing Genome : Escherichia_coli_ASM584v2\n",
      "Processing Chromosome : U00096.3\n",
      "Saving Chromosome : U00096.3\n",
      "Processing Genome : GCA_006094375.1_ASM609437v1_genomic\n",
      "Processing Chromosome : CP035288.1\n",
      "Saving Chromosome : CP035288.1\n",
      "Processing Chromosome : CP035289.1\n",
      "Saving Chromosome : CP035289.1\n",
      "Processing Chromosome : CP035290.1\n",
      "Saving Chromosome : CP035290.1\n",
      "Processing Genome : GCF_022354085.1_ASM2235408v1_genomic\n",
      "Processing Chromosome : NZ_CP055055.1\n",
      "Saving Chromosome : NZ_CP055055.1\n",
      "Processing Chromosome : NZ_CP055056.1\n",
      "Saving Chromosome : NZ_CP055056.1\n",
      "CPU times: total: 10min 56s\n",
      "Wall time: 6min 18s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "def Master_Interpolator_Generation(Genome_FNA_path, Kmer_size, Num_chunks, Output_dir): \n",
    "    \"\"\"\n",
    "    Takes a raw FNA file, and creates a .pkl file.\n",
    "    Pkl contains the equations for getting the expected \"mean_kmer_count\",\n",
    "    and the expected \"Cov_matrix\".\n",
    "    Creates a seperate file for each kmer_size \n",
    "    \"\"\"\n",
    "    # Get the Genome name from the Genome_FNA_path \n",
    "    # Chromosome.description appears too variable \n",
    "\n",
    "    Genome_name_ext = Path(Genome_FNA_path).name\n",
    "\n",
    "    Genome_name = \".\".join(Genome_name_ext.split(\".\")[:-1])\n",
    "\n",
    "    # Builds a fast look up system for finding the canonical kmer (lexographical)\n",
    "    Canonical_map = Build_canonical_map(Kmer_size) \n",
    "\n",
    "    print(f\"Processing Genome : {Genome_name}\")\n",
    "    for Chromosome in SeqIO.parse(Genome_FNA_path, \"fasta\"):\n",
    "        print(f\"Processing Chromosome : {Chromosome.id}\")\n",
    "        # A list (in order of smallest to largest) of valid chunk_sizes for the given chromosome\n",
    "        Valid_chunk_sizes = Get_valid_chunk_sizes(len(Chromosome.seq))\n",
    "        \n",
    "        # returns an np.array of the mean kmer counts across all chunk sizes (shape of : n_chunk_sizes, n_kmers). \n",
    "        #And the associated covariance matrix (shape of : n_chunk_sizes, n_kmers, n_kmers)\n",
    "        Mean_counts_list, Mean_cov_list = Generate_arrays(Valid_chunk_sizes, Kmer_size, Canonical_map, Chromosome.seq, Chromosome.id, Num_chunks)\n",
    "    \n",
    "        # A list of interpolator functions for each kmer in the Mean_counts_array\n",
    "        Mean_counts_interpolator = Generate_counts_interpolator(Mean_counts_list, Valid_chunk_sizes)\n",
    "        # A matrix of interpolator functins for each pair of kmers in the mean_cov_array \n",
    "        Cov_interpolator = Generate_cov_interpolator(Mean_cov_list, Valid_chunk_sizes)\n",
    "\n",
    "        # Save the interpolators\n",
    "        print(f\"Saving Chromosome : {Chromosome.id}\")\n",
    "        Export_Interpolators(Mean_counts_interpolator, Cov_interpolator, Valid_chunk_sizes, Kmer_size, Chromosome.id, Genome_name, Output_dir)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Num_chunks = 150\n",
    "Kmer_size = 4\n",
    "Output_dir = \"C:/users/henry chapman/Documents/Galaxy/Output/Pipe_1/Interpolators/Kmer_size_4/Raw_counts\"\n",
    "\n",
    "for Genome_FNA_path in Path(\"C:/users/henry chapman/Documents/Galaxy/Raw_data/Genomes\").iterdir():\n",
    "    if \".ipynb_checkpoints\" in Genome_FNA_path.name:\n",
    "        continue\n",
    "    Master_Interpolator_Generation(Genome_FNA_path, \n",
    "                               Kmer_size, \n",
    "                               Num_chunks, \n",
    "                               Output_dir)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3410e5e8-c584-4f26-a5f6-aa08ba299b8e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d8b7c04-f9e8-41d8-8373-dbe77045b4e5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8051d24-fc23-4094-818c-be9df6cedda1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
